src/extensions/optionalFieldsDirectiveExtension.ts

```diff
@@ -0,0 +1,32 @@
+import { getOptionalFieldsDirectivePaths } from "lib/getOptionalFieldsDirectivePaths"
+import { isEqual } from "lodash"
+import { flattenErrors, statusCodeForError } from "lib/graphqlErrorHandler"
+
+export const optionalFieldsDirectiveExtension = (documentAST, result) => {
+  const paths = getOptionalFieldsDirectivePaths(documentAST)
+  const extensions = {}
+  const optionalFields: any[] = []
+
+  if (paths.length && result.errors && result.errors.length) {
+    const optionalErrors = result.errors.filter(
+      (error) => paths.filter((path) => isEqual(error.path, path)).length
```

Comment:
I wonder if errors that happen somewhere in a subtree of an optional field (annotated with @optionalField) should be included in the optionalField extension as well 🤔 

For example, fetching the `review` could fail in the following query. With our current implementation, the error wouldn't show up in the `optionalFields` extension because we check if the error has the exact same path as the `optionalField` directive.

```
query { 
  artwork(id: "an-id") @optionalField {
    title
    review {
       text
    }
  }
}
```

I guess we have two options:

1. Don't include errors from subtree: This could lead to some "Unable to load" screens because our Eigen error middleware would throw the error if it is not included in the `optionalFields` extension.

2. Include errors from subtree: Every component rendering a field annotated with `@optionalField` would need to handle null values everywhere in the subtree.

What do you think @brainbicycle?

Reply:
This would be the implementation for Option 2.

```suggestion
      (error) => paths.filter((path) => isEqual(error.path.slice(0, path.length), path))
```

Reply:
I think option 2 makes the most sense to me

Reply:
I agree, let's go with option 2 then!

Reply:
Yeah, option 2 looks more intuitive.

I just want to mention that there could be cases where a component ends up in an "incomplete" state. For example, letting fetching the `title` fail in the following query could end up in a list of articles with missing titles.

```
query { 
  articles @optionalField {
    title
    ...
  }
}
```

![simulator_screenshot_8F42235B-7221-4BBE-BA9F-A5E2E96B8F04](https://user-images.githubusercontent.com/4691889/123631990-b0fab580-d817-11eb-9938-e314be3f4e92.png)

Although I guess that's fine because in practice the example above (fetching reviews fails) is more realistic and I can't find an example where just a single field with a basic type could fail.